product_abc_classification/models/abc_classification_profile.py

```diff
@@ -140,17 +140,39 @@ def _get_sort_key_percentage(rec):
             )
             percentages = levels.mapped("percentage")
             level_percentage = list(zip(levels, percentages))
+            current_value = 0
+            accumulated_percentage = level_percentage[0][1]
             for product_data in product_list:
-                product_data["value_percentage"] = (
-                    (100.0 * product_data["value"] / totals["value"])
-                    if totals["value"]
-                    else 0.0
-                )
+                product_data["product"].abc_classification_level_id = level_percentage[
+                    0
+                ][0]
+                # Accumulated current value
+                current_value += product_data["value"] or 0.0
+                # This comparison would be the same as:
+                # current_value * 100 / totals["value"] > accumulated_percentage,
+                # but it is written in the next way to avoid division and decimal lost.
                 while (
-                    product_data["value_percentage"] < level_percentage[0][1]
+                    current_value * 100 > accumulated_percentage * totals["value"]
                     and len(level_percentage) > 1
                 ):
+                    # This condition is done in cases where the previous value or/and
+                    # the current value are high.
+
+                    # In this case, that product would be assigned the next level.
+                    # Example: if the levels are 80%/15%/5% and the previous value
+                    # was 79% and the current value 94%,
+                    # without this condition this product would be assigned the level A,
+                    # but with this condition the level B, which would be more accurate.
+
+                    # The first part of the condition is just the mean of the previous and
+                    # current value, but written in this way:
+                    # (Pv + Cv)/2  = Cv - (Cv - Pv)/2.
+                    # Where Cv- Pv is the value of the product.
+                    if (
+                        current_value - (product_data["value"] or 0.0) / 2
+                    ) * 100 > accumulated_percentage * totals["value"]:
+                        product_data[
+                            "product"
+                        ].abc_classification_level_id = level_percentage[1][0]
```

Comment:
```suggestion
```
this is not needed if you apply the other two changes.

Reply:
In the case of the easystock your code and mine do the exact classification, you are right. But this extra code that I added was thought for the case where a product had an accumulated percentage slightly over the threshold. So maybe, it would be better classified at the previous level. For example, imagine this situation: 
![Captura de pantalla de 2022-02-02 09-56-43](https://user-images.githubusercontent.com/79082305/152122919-ef22c9e8-3ed7-40fa-b450-be46eb3a2ead.png)
As you can see the results are a little different. However, I agree with you that your solution is better because in a real database with a lot of products, the difference between both codes is insignificant and there is no need to complicate it that much.

Reply:
Oh, thanks for your explanation and comprehension :)